From 9cc9c0a3788ed165ba1efeac2543e9162f20325d Mon Sep 17 00:00:00 2001
From: Lokesh Vutla <lokeshvutla@ti.com>
Date: Tue, 20 Oct 2020 16:07:40 +0530
Subject: [PATCH 03/26] firmware: ti_sci: Use dev_id as resource type with ABI
 3.0

With ABI 3.0, sysfw deprecated special resource types used for AM65x
SoC. So, make ti_sci_get_resource_type() to return resource type as
devid if ABI 3.0 is detected.

Signed-off-by: Lokesh Vutla <lokeshvutla@ti.com>
---
 drivers/firmware/ti_sci.c              | 17 ++++++++++++++---
 include/linux/soc/ti/ti_sci_protocol.h |  6 ++++++
 2 files changed, 20 insertions(+), 3 deletions(-)

diff --git a/drivers/firmware/ti_sci.c b/drivers/firmware/ti_sci.c
index 1b73e7b90d28..4b21139f5195 100644
--- a/drivers/firmware/ti_sci.c
+++ b/drivers/firmware/ti_sci.c
@@ -1729,6 +1729,15 @@ static int ti_sci_cmd_core_reboot(const struct ti_sci_handle *handle)
 	return ret;
 }
 
+bool ti_sci_abi_3_and_above(const struct ti_sci_handle *handle)
+{
+	if (handle->version.abi_major >= 3)
+		return true;
+	else
+		return false;
+}
+EXPORT_SYMBOL_GPL(ti_sci_abi_3_and_above);
+
 static int ti_sci_get_resource_type(struct ti_sci_info *info, u16 dev_id,
 				    u16 *type)
 {
@@ -1736,8 +1745,11 @@ static int ti_sci_get_resource_type(struct ti_sci_info *info, u16 dev_id,
 	bool found = false;
 	int i;
 
-	/* If map is not provided then assume dev_id is used as type */
-	if (!rm_type_map) {
+	/*
+	 * - If map is not provided then assume dev_id is used as type.
+	 * - With abi 3.0, sysfw depricated special resource types. use dev_id as type.
+	 */
+	if (!rm_type_map || ti_sci_abi_3_and_above(&info->handle)) {
 		*type = dev_id;
 		return 0;
 	}
@@ -3246,7 +3258,6 @@ devm_ti_sci_get_resource_sets(const struct ti_sci_handle *handle,
 			      struct device *dev, u32 dev_id, u32 *sub_types,
 			      u32 sets)
 {
-	u32 resource_subtype;
 	u16 resource_type;
 	struct ti_sci_resource *res;
 	bool valid_set = false;
diff --git a/include/linux/soc/ti/ti_sci_protocol.h b/include/linux/soc/ti/ti_sci_protocol.h
index c05502bf337a..1adb93d5ca03 100644
--- a/include/linux/soc/ti/ti_sci_protocol.h
+++ b/include/linux/soc/ti/ti_sci_protocol.h
@@ -562,6 +562,7 @@ struct ti_sci_resource *
 devm_ti_sci_get_resource(const struct ti_sci_handle *handle, struct device *dev,
 			 u32 dev_id, u32 sub_type);
 
+bool ti_sci_abi_3_and_above(const struct ti_sci_handle *handle);
 #else	/* CONFIG_TI_SCI_PROTOCOL */
 
 static inline const struct ti_sci_handle *ti_sci_get_handle(struct device *dev)
@@ -616,6 +617,11 @@ devm_ti_sci_get_resource(const struct ti_sci_handle *handle, struct device *dev,
 {
 	return ERR_PTR(-EINVAL);
 }
+
+static inline bool ti_sci_abi_3_and_above(const struct ti_sci_handle *handle)
+{
+	return false;
+}
 #endif	/* CONFIG_TI_SCI_PROTOCOL */
 
 #endif	/* __TISCI_PROTOCOL_H */
-- 
2.31.1

